### PR TITLE
[spark] Dynamically adjust the parallelism of scan

### PR DESCRIPTION
**Pull request by YannByron**
### Purpose

Linked issue: https://github.com/apache/incubator-paimon/issues/2404

### CODE REVIEW

**Comment by YannByron** [comment]
This optimization is similar to `FilePartition.getFilePartitions` in DataSource V1, which adjust the parallelism based on available cores.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/ScanHelper.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        case _ => false
+      }
+      val reshuffled = reshuffleSplits0(toReshuffle.collect { case ds: DataSplit => ds })
+      reshuffled ++ reserved
+    } else {
+      splits
+    }
+  }
+
+  private def reshuffleSplits0(splits: Array[DataSplit]): Array[DataSplit] = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/ScanHelper.scala:56]: Can you add a test for this method?

Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/ScanHelper.scala:56]: OK~

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**